lib/url/UrlJavascriptGenerator.js

```diff
@@ -0,0 +1,85 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Yuta Hiroto @hiroppy
+*/
+
+"use strict";
+
+const { RawSource } = require("webpack-sources");
+const Generator = require("../Generator");
+const RuntimeGlobals = require("../RuntimeGlobals");
+
+/** @typedef {import("webpack-sources").Source} Source */
+/** @typedef {import("../Compilation")} Compilation */
+/** @typedef {import("../Compiler")} Compiler */
+/** @typedef {import("../Generator").GenerateContext} GenerateContext */
+/** @typedef {import("../NormalModule")} NormalModule */
+/** @typedef {import("../RuntimeTemplate")} RuntimeTemplate */
+
+const TYPES = new Set(["url"]);
+
+class UrlJavascriptGenerator extends Generator {
+	/**
+	 * @returns {Set<string>} available types (do not mutate)
+	 */
+	getTypes() {
+		return TYPES;
+	}
+
+	/**
+	 * @param {Compilation} compilation the compilation
+	 */
+	constructor(compilation) {
+		super();
+		this.mainTemplate = compilation.mainTemplate;
+	}
+
+	/**
+	 * @param {NormalModule} module module for which the code should be generated
+	 * @param {GenerateContext} generateContext context for generate
+	 * @returns {Source} generated code
+	 */
+	generate(module, { chunkGraph, runtimeTemplate, runtimeRequirements }) {
+		const originalSource = module.originalSource();
+
+		if (!originalSource) {
+			return new RawSource("throw new Error('No source available');");
+		}
+
+		const filename = module.resource;
+		const { urlModuleFilename } = runtimeTemplate.outputOptions;
+		const url = this.mainTemplate.getAssetPath(urlModuleFilename, {
+			module,
+			filename,
+			chunkGraph
+		});
+
+		runtimeRequirements.add(RuntimeGlobals.module);
+		runtimeRequirements.add(RuntimeGlobals.exports);
+		runtimeRequirements.add(RuntimeGlobals.require);
```

Comment:
```suggestion
```

Not needed as you don't use the `__webpack_require__` variable (directly).

The fact that `RuntimeGlobals.publicPath` depends on `__webpack_require__` is handled somewhere else (RuntimePlugin).